tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc

```diff
@@ -117,51 +339,238 @@ void TRTEngineOp::Compute(OpKernelContext* context) {
+      return engine_map.at(static_engine->getMaxBatchSize());
+    }
+  } else {
+    auto engine_it = engine_map.find(batch_size);
+    if (engine_it == engine_map.end() &&
+        engine_map.size() < (size_t)max_cached_engines) {
+      auto builder_ = std::shared_ptr<nvinfer1::IBuilder>(
+          nvinfer1::createInferBuilder(logger),
+          Destroyer<nvinfer1::IBuilder>());  // reset the builder to ensure
+                                             // device is correct
+#if NV_TENSORRT_MAJOR > 3
+      auto device = context->device();
+      auto device_name = device->name();
+      if (allocators_.count(device_name)) {
+        builder_->setGpuAllocator(allocators_.at(device_name).get());
+      } else {
+        std::make_shared<TRTDeviceAllocator> auto dev_allocator =
+            device->GetAllocator(tensorflow::AllocatorAttributes());
+        if (!dev_allocator) {
+          LOG(ERROR) << "Can't find device allocator for gpu device "
+                     << device->name();
+          ctx->SetStatus(
+              tensorflow::errors::Internal("Can't get device allocator"));
+          return nullptr;
+        }
+        auto allocator_ = std::make_shared<TRTDeviceAllocator>(dev_allocator);
+        builder_->setGpuAllocator(allocator_.get());
+        allocators_.insert({device_name, allocator});
+      }
+#endif
+      VLOG(1) << name() << " Constructing a new engine with batch size "
+              << batch_size;
+      builder_->setMaxBatchSize(batch_size);
+      if (precision_mode == tensorflow::tensorrt::convert::FP16MODE) {
+        builder_->setHalf2Mode(true);
+      } else if (precision_mode == tensorflow::tensorrt::convert::INT8MODE) {
+        builder_->setInt8Mode(true);
+        builder_->setInt8Calibrator(calibrator_.get());
+      }
+      builder_->setMaxWorkspaceSize(workspace_size_);
+      nvinfer1::ICudaEngine* engine = nullptr;
+      std::vector<tensorflow::PartialTensorShape> shapes;
+      for (int i = 0; i < ctx->num_inputs(); ++i) {
+        shapes.emplace_back(ctx->input(i).shape());
+      }
+      auto status = tensorflow::tensorrt::convert::ConvertSubgraphToEngine(
+          segment_graph_, builder_.get(), shapes, &engine, precision_mode);
+      if (engine) {
+        engine_map[batch_size] = {
+            std::shared_ptr<nvinfer1::ICudaEngine>(
+                engine, Destroyer<nvinfer1::ICudaEngine>()),
+            std::shared_ptr<nvinfer1::IExecutionContext>(
+                engine->createExecutionContext(),
+                Destroyer<nvinfer1::IExecutionContext>())};
+      } else {
+        LOG(ERROR) << "Engine creation for batch size " << batch_size
+                   << " failed";
+        ctx->SetStatus(tensorflow::errors::Internal("Engine creation failed!"));
+        engine_map[batch_size] = {nullptr, nullptr};
+        return {nullptr, nullptr};
+      }
+    }
+    return engine_map.at(batch_size);
+  }
+}
+
+tensorflow::Status TRTEngineOp::AllocateCalibrationResources(
```

Comment:
Is this purely copied from TrtCalibOp? Or any diffs?

Reply:
CalibOp is not there anymore

Reply:
Yes, I know, I just wanted to make sure the logic is same (so I can stop reviewing it)

Reply:
Yes, it is mostly copy-paste with small modifications